### ISSUE TITLE
Google maps markers with custom icons look big on iOS

### ISSUE THREAD

**Issue by arnoraps** [flutter/flutter#24865]
When using custom icons on the markers (32 x 32 pixel size, png format) the icons look big on iOS, but very small on Android devices.
![iconssmall](https://user-images.githubusercontent.com/33053001/49243318-573a7080-f40d-11e8-9627-3c763d029d21.jpg)
![largeicons](https://user-images.githubusercontent.com/33053001/49243319-57d30700-f40d-11e8-88e3-1e4f497e43b5.jpg)

`MarkerOptions(                              
            position: LatLng(
              ..
            ),
            icon: BitmapDescriptor.fromAsset(crumb.marker),`

**Comment by ZeroCool00** [comment]
@arnoraps could you show me how did you add multiple markers on map?

**Comment by arnoraps** [comment]
`  void _onMapCreated(GoogleMapController controller) {
    setState(() { mapController = controller; });

    for(var i = 0; i < array.length; i++) {
      var item = array[i];
      if (item!=null && item.gps!=null) {
        controller.addMarker(        
          MarkerOptions(                              
            position: LatLng(
              item.gps.lat,
              item.gps.lon,
            ),
            icon: BitmapDescriptor.fromAsset(item.marker),
          ),
        );   
      }
    }
  }`

**Comment by yohom** [comment]
This is on iOS:
![iOS](https://user-images.githubusercontent.com/10418364/49432408-93454b00-f7ea-11e8-805e-eb794a7c6823.png)
This is on Android:
![Android](https://user-images.githubusercontent.com/10418364/49432415-96d8d200-f7ea-11e8-9dbc-765d523ddbc1.jpg)

**Comment by ZeroCool00** [comment]
@arnoraps  thank you so much
@zoechi same issue on ios, icon is too big. android is good.

**Comment by DanialV** [comment]
why there is no option to control marker icon size?

**Comment by cqhchan** [comment]
is there any updates on this?

**Comment by giandifra** [comment]
can i insert custom widget like icon ? for aggregation point i need to write the number of pin into icon of marker....
thank you

**Comment by giandifra** [comment]
@arnoraps thanks

**Comment by sam-toon** [comment]
I have just been playing with this.  I was supplying image files at higher resolution, @2x and @3x for iOS and the equivalent for Android.  These files are just being ignored and only the base resolution image is being used. iOS then automatically scales the images for higher resolution screens,  but Android doesn't do this automatic scaling and just displays the icon at 1x.  This is why the icons appear larger on iOS than on Android.

**Comment by Toolenaar** [comment]
Having the same problems. Icons on iOS are too big and pixelated.

**Comment by cqhchan** [comment]
this is a temporary fix if anyone needs

    BitmapDescriptor get deliveryIcon {
      bool isIOS = Theme.of(context).platform == TargetPlatform.iOS;
          If (isIOS)
              return BitmapDescriptor.fromAsset('assets/icons/orange_pin.png');
         else
              return BitmapDescriptor.fromAsset(
              'assets/icons/3.0x/orange_pin.png');
    }


Simply put, supply the android the larger asset.

**Comment by marti1125** [comment]
guys Is it possible when you click in marker go to another page ?

**Comment by arnoraps** [comment]
> guys Is it possible when you click in marker go to another page ?

Yes, but you to use a trick to detect which marker it is (see other reported issue with no custom id available)

**Comment by m7amdElberawy** [comment]
@arnoraps Could you show me how to grouping markers with there markers' count ??

**Comment by arnoraps** [comment]
> @arnoraps Could you show me how to grouping markers with there markers' count ??

I haven't done that. The numbers on my icons were generated for the icons, not for grouping

**Comment by kevinApplaud** [comment]
HI, is this issue resolved now?

**Comment by sam-toon** [comment]
I have had another look at this, it hasn't been fixed.

To get iOS to work properly and use the @2x and @3x images, you need to add the image variants to the pubspec.yaml file, so you need:
````
assets:
  - assets/images/icon.png
  - assets/images/icon@2x.png
  - assets/images/icon@3x.png
````
otherwise the files don't get copied to the iOS build folder.

Then something like the following code will return the correct filename to use for the icon.

````dart
iconFile() {
  if (Platform.isAndroid) {
    double mq = MediaQuery.of(context).devicePixelRatio;
    String dirName = ''; // default for 1.0x
    if (mq > 1.5 && mq < 2.5) {
      dirName = "2.0x/";
    } else if (mq >= 2.5) {
      dirName = "3.0x/";
    }
    return 'assets/images/' + dirName + 'icon.png';
  }
  // this is for iOS
  return 'assets/images/icon.png';
}
````

**Comment by kuprel** [comment]
The small icons are blown up on iOS and look fuzzy/pixelated

**Comment by juanjoserodrigolazaro** [comment]
any update on this, having the same issue!

**Comment by zoechi** [comment]
@juanjoserodrigolazaro 
Using "add reaction" on the initial comment would increase priority while `+1` comments are rather pointless and cause lots of people to get notified for no reason.
To get notified yourself use the [Subscribe] button to the top right instead.

**Comment by collinjackson** [comment]
It looks like flutter/plugins#815 might provide an acceptable workaround for this issue

**Comment by abumoallim** [comment]
> guys Is it possible when you click in marker go to another page ?

There is OnTap method which you can use

**Comment by gtupak** [comment]
> It looks like [flutter/plugins#815](https://github.com/flutter/plugins/pull/815) might provide an acceptable workaround for this issue

For me, the icons are still much bigger and fuzzy, where the Android ones are fine. 

> It looks like [flutter/plugins#815](https://github.com/flutter/plugins/pull/815) might provide an acceptable workaround for this issue

I tried it and I don't see any difference.

@arnoraps, did you find a fix yet?

**Comment by collinjackson** [comment]
Based on @gtupak's comment it looks like there may still be work to do here to get the icons on iOS to look as expected. I am not actively working on it right now, so unassigning myself. If anyone wants to open a PR for a fix, I'm happy to review that.

/cc @iskakaushik

**Comment by iskakaushik** [comment]
@gtupak , can you try version `0.5.7` and use `BitmapDescriptor.fromAssetImage` and let me know if that solves this?

**Comment by gtupak** [comment]
Wow, great work, guys, the icons are FINALLY not pixelated. Big cheers! 🙌

**Comment by sam-toon** [comment]
Fantastic! The icons are now scaled correctly on both iOS and Android.

**Comment by iskakaushik** [comment]
https://github.com/flutter/plugins/pull/1478/files fixes this issue.

**Comment by sky93** [comment]
شادی باشه

**Comment by DanialV** [comment]
Wow, such a great update finally I am waiting for this PR and polyline PR for about 3 months now finally they both available thanks so much.

**Comment by workerbee22** [comment]
So BitmapDescriptor.fromAssetImage returns a Future<BitmapDescriptor> ... so do you need to then use a FutureBuilder to render the Markers? How do I complete the Future to give to the Marker icon property?

**Comment by aytunch** [comment]
Guys and gals, does this allow for a custom widget to be used as a marker?

**Comment by joserocha3** [comment]
@workerbee22 check that example. They store the final marker icon value into state.

https://github.com/flutter/plugins/blob/master/packages/google_maps_flutter/example/lib/marker_icons.dart

**Comment by ollydixon** [comment]
Would be nice if we could just have a widget for many reasons.. anyhow new bug.

`BitmapDescriptor.fromAssetImage` ignores the image size as well. Example loading: 

```
 BitmapDescriptor.fromAssetImage(
            ImageConfiguration(size: Size(60, 60)), 'assets/meetup_active.png')
        .then((onValue) {
      setState(() {
        markerIcon = onValue;
      });
    });
```

**Comment by th3g3ntl3m3n** [comment]
Custom image Size is not working in 0.5.13.

**Comment by duzenko** [comment]
Does not work for me with single asset images (no 2.0, 3.0, etc).
I would expect this to 'just work' and scale marker icons automatically - like it does on Android.

**Comment by d3xt3r2909** [comment]
I is not working also in 0.5.16

**Comment by GregorySech** [comment]
@iskakaushik  can this issue be re-opened please? As stated before the size parameter is ignored.

**Comment by iskakaushik** [comment]
@GregorySech This is the intended behavior. The `size` parameter in the `ImageConfiguration` is ignored. We only use `ImageConfiguration` to resolve to the image of the correct `dpi` scaled resolution, `2x`, `3x` etc.

I have created a new issue to address this: https://github.com/flutter/flutter/issues/34657, these add a thumbs-up to it. It will help us prioritize.

**Comment by ollydixon** [comment]
> @GregorySech This is the intended behavior. The `size` parameter in the `ImageConfiguration` is ignored. We only use `ImageConfiguration` to resolve to the image of the correct `dpi` scaled resolution, `2x`, `3x` etc.
> 
> I have created a new issue to address this: #34657, these add a thumbs-up to it. It will help us prioritize.

You need to remove this though. Replace it with a Widget. The whole ImageConfiguration just feels unnecessary and not well thought-out.

**Comment by iskakaushik** [comment]
@ollydixon , there is an issue that tracks it https://github.com/flutter/flutter/issues/24213. It's an involved change as it requires us to synchronously keep the marker positions and the google map state in sync. If you look at the way we currently communicate with the platforms it's through out platform channels api, which doesn't yet allow synchronous communication.

We added support for markers to be created from:

- raw image bytes
- network images
- asset images

I agree that these don't solve all the use-cases but the marker-widget issue is not part of our [roadmap](https://github.com/flutter/flutter/wiki/Roadmap) as it requires some additional work to be unblocked.

**Comment by ollydixon** [comment]
@iskakaushik wouldn't that just be widgets with coordinates, google maps API gives you all the points with lat and long? The bullet points you listed could easily be loaded from special marker widgets. It's how they do it on the React Google maps implementation; you really have no excuse sorry. One of the main philosophies of Flutter is 'everything is a widget' you're breaking it with this strange API; should be no.1 on your roadmap; you are losing customers to Flutter MapBox which is far superior at the moment.

**Comment by almohitapp** [comment]
Simple way to solve this problem by using that

BitmapDescriptor get deliveryIcon {
  bool isIOS = Theme.of(context).platform == TargetPlatform.iOS;
  if (isIOS)
    return BitmapDescriptor.fromAsset('assets/icons/orange_pin.png');
  else
    return BitmapDescriptor.fromAsset(
        'assets/icons/3.0x/orange_pin.png');
}

**Comment by oodavid** [comment]
@ollydixon GoogleMaps aren't a typical flutter widget - they're actually an embedded Android / iOS view (ala React-Native). It's unlikely to have Widget icons before a major refactor.

> The plugin relies on Flutter's new mechanism for embedding Android and iOS views. As that mechanism is currently in a developers preview, this plugin should also be considered a developers preview. [~GoogleMaps on pub.dev](https://pub.dev/packages/google_maps_flutter)

**Comment by ollydixon** [comment]
@oodavid thanks for the info, you could still use a stack-view in this circumstance though. As long as you have the markers meta info. Just an idea though. :-)

**Comment by timeuser** [comment]
I upgraded to Flutter 1.7 and now this issue is back. Using google_maps_flutter 0.5.19+2 with Flutter 1.7.8+hotfix.3 from stable channel. BitmapDescriptor.fromAssetImage() is using the 1.0x asset for devices with higher resolution so the marker images are tiny.

**Comment by ollydixon** [comment]
@timeuser may I suggest using using https://github.com/mapbox/flutter-mapbox-gl it's much better. Uses widgets, customizable maps and it's cheaper.

**Comment by duzenko** [comment]
> 
> 
> I upgraded to Flutter 1.7 and now this issue is back. Using google_maps_flutter 0.5.19+2 with Flutter 1.7.8+hotfix.3 from stable channel. BitmapDescriptor.fromAssetImage() is using the 1.0x asset for devices with higher resolution so the marker images are tiny.

https://github.com/flutter/plugins/pull/1821

**Comment by KingWu** [comment]
i am using Flutter 1.5.4 and google_maps_flutter 0.5.19+2. Still load 1.0x asset

**Comment by harshaIOT** [comment]
@timeuser  is right this issue is back again please suggest us what to do??

**Comment by rlechetaudemy** [comment]
Why it's so complicated to change the marker icon? The goals of use Flutter should be to make apps development easier, right?
Are you talking serious about doing all this code just to change the icon? 
https://github.com/flutter/plugins/blob/master/packages/google_maps_flutter/example/lib/marker_icons.dart

**Comment by duzenko** [comment]
> 
> 
> Why it's so complicated to change the marker icon? The goals of use Flutter should be to make apps development easier, right?
> Are you talking serious about doing all this code just to change the icon?
> https://github.com/flutter/plugins/blob/master/packages/google_maps_flutter/example/lib/marker_icons.dart

I agree. I proposed a contextless, sync version of  BitmapDescriptor.fromAssetImage but they ignored me.

**Comment by ollydixon** [comment]
The React native implmentation has markers as components. Example:

```
<MapView
  region={this.state.region}
  onRegionChange={this.onRegionChange}
>
  {this.state.markers.map(marker => (
    <Marker
      coordinate={marker.latlng}
      title={marker.title}
      description={marker.description}
    />
  ))}
</MapView>
```

I'm sorry contributers this is one of the worst ports of Google Maps I've ever seen...

@duzenko` BitmapDescriptor.fromAssetImage` shouldn't even exist in the first place.

**Comment by DKMonster** [comment]
Try this, It's work for me.

https://stackoverflow.com/questions/53633404/how-to-change-the-icon-size-of-google-map-marker-in-flutter

**Comment by IdrisN** [comment]
> Why it's so complicated to change the marker icon? The goals of use Flutter should be to make apps development easier, right?
> Are you talking serious about doing all this code just to change the icon?
> https://github.com/flutter/plugins/blob/master/packages/google_maps_flutter/example/lib/marker_icons.dart

I tried this cumbersome method but it still does not work consistently for android and ios. Drawing your own canvas seems to work but if the marker gets to complicated, it's not very easy to draw it since you cannot import a path from a svg. ..Sigh.

**Comment by thalkz** [comment]
Not really a great fix, but this solved it for me.

In the `google_maps_flutter` plugin, update the file `lib/src/bitmap.dart` like this :

```
static Future<BitmapDescriptor> fromAssetImage(...) {
    ...
    return BitmapDescriptor._(<dynamic>[
      'fromAssetImage',
      assetBundleImageKey.name,
      4.0, // <= Replace 'assetBundleImageKey.scale' by '4.0' on line 76
    ]);
}
```

Icon markers are no longer are too big or blurry on iOS. You might want to add a check for Platform.isIOS to only apply the fix for iOS.

I hope it helps !

**Comment by DoubleHub** [comment]
> Not really a great fix, but this solved it for me.
> 
> In the `google_maps_flutter` plugin, update the file `lib/src/bitmap.dart` like this :
> 
> ```
> static Future<BitmapDescriptor> fromAssetImage(...) {
>     ...
>     return BitmapDescriptor._(<dynamic>[
>       'fromAssetImage',
>       assetBundleImageKey.name,
>       4.0, // <= Replace 'assetBundleImageKey.scale' by '4.0' on line 76
>     ]);
> }
> ```
> 
> Icon markers are no longer are too big or blurry on iOS. You might want to add a check for Platform.isIOS to only apply the fix for iOS.
> 
> I hope it helps !

I do not recommend modifying the library code directly, as it would only work on your personal installation of the plugin. If you try to clean cache / reinstall plugins for some reason, or even build the project on another PC or with a CI tool, it won't work anymore (of course). So take this path only if you really know what you are doing

**Comment by Toolenaar** [comment]
Why is this issue closed ? I Still have this issue. On iOS the icons are tiny.

**Comment by duzenko** [comment]
> 
> 
> Why is this issue closed ? I Still have this issue. On iOS the icons are tiny.

https://github.com/flutter/plugins/pull/1821

**Comment by rafaelferreirapt** [comment]
Why is this issue closed ? I Still have this issue. On iOS the icons are tiny.

**Comment by Marek00Malik** [comment]
What is the status of this issue? This should be priorities and not thrown to the bottom of the stack.. it's 1.5 year since this ticket was opened.
